kong/plugins/apifortress/fortress_serializer.lua

```diff
@@ -0,0 +1,117 @@
+-- The serializer. Builds a data structure that is compatible with
+-- the API Fortress API expectations
+
+local stringy = require "stringy"
+
+local _M = {}
+
+-- pre-indexing global functions
+local str_startswith = stringy.startswith
+local str_endswith = stringy.endswith
+local str_sub = string.sub
+local str_len = string.len
+local table_insert = table.insert
+local encode_base64 = ngx.encode_base64
+
+function _M.serialize(ngx)
+
+  -- pre-indexing to reduce the number of lookups later on
+  local ngx_req = ngx.req
+  local ngx_resp = ngx.resp
+  local ngx_var = ngx.var
+  local ngx_ctx = ngx.ctx
```

Comment:
Just realized this is in the `serialize` function. Kind of lost in all those comments, I didn't see the line above, sorry about that, `ctx` caching is fine here yes.

However, instead of caching, say, `ngx.req`, it is more efficient to cache at the module level, say, `local get_body_data = ngx.req.get_body_data`. Same goes for most of them here.